Attend from pivot to model

david-fairbanks42 opened the pull request:
Inspired by http://transmission.vehikl.com/theres-a-model-hiding-in-your-rest-api/ I've changed the attends pivot to a full model. It's a large change but I think it adds some clarity over using just the pivot table.
I've also added API endpoints for meetings and attendance. There may be refactoring required with the URIs for the endpoints depending on the style we wish to use.

david-fairbanks42 committed:
Change meeting attendees from pivot table to model

david-fairbanks42 committed:
Flush out meeting and attend API endpoints

Includes unit tests for handlers and feature tests for request endpoints.

Renamed and refactored Attendee to Attend since it is the record of the users attend status to a meeting.

david-fairbanks42 committed:
Merge branch 'master' into attend-from-pivot-to-model

david-fairbanks42 committed:
Attend search after Meetup discussion

david-fairbanks42 committed:
Remove handlers and add meeting policy

Move meeting and attend validation rules to form requests.
Converted meeting search to a query scope under meeting.
Added more negative feature tests.
Changed composer PHP requirement to 7.4 to allow arrow function.

david-fairbanks42 committed:
Reformat routes to be consistent with project

david-fairbanks42 committed:
Rename Attend records to Reservation records

FatBoyXPC commented on app/Http/Controllers/Api/MeetingController.php:
```diff
@@ -0,0 +1,121 @@
+    {
+        foreach($request->validated() as $field => $value) {
+            $meeting->{$field} = $value;
+        }
```
Can we instead `$meeting->update($request->validated())`? I'm not a big fan of dynamic variable access.

FatBoyXPC commented on app/Http/Controllers/Api/MeetingController.php:
```diff
@@ -0,0 +1,121 @@
+                    'message' => 'Error deleted meeting'
+                ]
+            );
+        }
```
I'm not a big fan of this if statement. It's a classic "defensive programming" technique but I'm fairly positive if that `delete()` fails then we'll have a bigger problem on our hands (such as an uncaught exception). Could we remove it?

FatBoyXPC commented on app/Http/Controllers/Api/MeetingController.php:
```diff
@@ -0,0 +1,121 @@
+     */
+    public function index(Request $request)
+    {
+        $meetings = Meeting::search($request->input());
```
I think you need to chain `->get()` onto the end of this? I believe the `count()` on L29 will be a call. I might be missing something, though!

david-fairbanks42 replied:
The count works without the get but you are correct, the response is different (more complete) with the get. Adding to test to catch difference.

FatBoyXPC commented on app/Http/Controllers/Api/ReservationController.php:
```diff
@@ -0,0 +1,171 @@
+            return response()->json(
+                [
+                    'success' => false,
+                    'message' => 'Error setting attending status'
```
Since this is an error case of some sort can we get an appropriate 4xx code into here?

FatBoyXPC commented on app/Http/Controllers/Api/ReservationController.php:
```diff
@@ -0,0 +1,171 @@
+            return response()->json(
+                [
+                    'success' => false,
+                    'message' => 'Error setting attending status'
```
Since this is an error case of some sort can we get an appropriate 4xx code into here?

FatBoyXPC commented on app/Http/Controllers/Api/ReservationController.php:
```diff
@@ -0,0 +1,171 @@
+     */
+    public function show(Meeting $meeting, User $user)
+    {
+        $attend = $user->reservations()->where(['meeting_id' => $meeting->getKey()])->first();
```
Can we use a `whereHas()` instead so we can avoid using the related_id field? It should be something like this:

```php
$user->reservations()->whereHas('meetings', function ($query) {
    $query->where($meeting->getKeyName(), $meeting->getKey());
});
```

I really dig what you were doing with `getKey()` to avoid specifically accessing `id`!

If you really wanted to, you could also do `$meeting->getTable()` instead of `'meetings'` in the `whereHas`.

FatBoyXPC commented on app/Http/Controllers/Api/ReservationController.php:
```diff
@@ -0,0 +1,171 @@
+    public function update(UpdateReservationRequest $request, Meeting $meeting)
+    {
+        /* @var Reservation $attend */
+        $attend = $request->user()->reservations()->where(['meeting_id' => $meeting->getKey()])->first();
```
Same thing as above about `whereHas`

FatBoyXPC commented on app/Http/Controllers/Api/ReservationController.php:
```diff
@@ -0,0 +1,171 @@
+    public function destroy(Request $request, Meeting $meeting)
+    {
+        /* @var Reservation $attend */
+        $attend = $request->user()->reservations()->where(['meeting_id' => $meeting->getKey()])->first();
```
Same thing as above about `whereHas`

FatBoyXPC commented on app/Http/Requests/StoreMeetingRequest.php:
```diff
@@ -0,0 +1,35 @@
+    public function authorize()
+    {
+        // should anyone be able to create a meeting or only users with specific roles?
+        return true;
```
The role should probably be the group's organizer(s), but I don't think we have a concept of that yet?

david-fairbanks42 replied:
I agree completely, that's why I left the note... I guess I would have normally used a `@todo`

FatBoyXPC commented on composer.json:
```diff
@@ -8,7 +8,9 @@
     "license": "MIT",
     "require": {
-        "php": "^7.2",
+        "php": "^7.4",
```
Good catch since we're using the new short arrow syntax!

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            [
+                'name' => $this->faker->company,
+                'description' => $this->faker->paragraph,
+                'location' => $this->faker->address,
```
I like putting actual values here instead of faker data. I've found that faker data can sometimes cause intermittent failures with weird characters.

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            ->assertStatus(201)
+            ->assertJson([
+                'success' => true
+            ]);
```
Can we also assert that the meeting gets created? We should be able to do something like `$messing = Meeting::first()` then `$this->assertEquals('Foo Name', $meeting->name)` etc.

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            ->assertJson([
+                'success' => true,
+                'rowCount' => 1
+            ]);
```
Can we also assert against the json containing the meeting?

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            ->assertJson([
+                'success' => true,
+                'rowCount' => 1
+            ]);
```
Can we test against the meetings showing up in the json?

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            [
+                'name' => 'Fake Meeting'
+            ]
+        );
```
Can we assert the meeting got updated? `$meeting = Meeting::first()` then `$this->assertEquals('Fake Meeting', $meeting->name)` etc.

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            ]);
+    }
+
+    public function test_update_onlyByOrganizer()
```
Can we snake case `only_by_organizer` as well?

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            ->assertStatus(403)
+            /*->assertJson([
+                'success' => false
+            ])*/;
```
Can we clean this assertion up?

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            ->assertStatus(200)
+            ->assertJson([
+                'success' => true
+            ]);
```
Can we also assert there are no meetings? `$this->assertEquals(0, Meeting::count())`

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            ]);
+    }
+
+    public function test_delete_onlyByOrganizer()
```
Can we snake case `only_by_organizer`?

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            ->assertStatus(403)
+            /*->assertJson([
+                'success' => false
+            ])*/;
```
Can we clean up this assertion and also assert the meeting is deleted?

FatBoyXPC commented on tests/Feature/MeetingTest.php:
```diff
@@ -0,0 +1,209 @@
+            ]);
+    }
+
+    public function test_search_notFound()
```
Can we snake case `not_found`?

FatBoyXPC commented on tests/Feature/ReservationTest.php:
```diff
@@ -0,0 +1,152 @@
+
+        $response
+            ->assertStatus(200)
+            ->assertJson(['success' => true]);
```
Can we also assert the reservation gets created?

FatBoyXPC commented on tests/Feature/ReservationTest.php:
```diff
@@ -0,0 +1,152 @@
+            ->assertJson([
+                'success' => true,
+                'rowCount' => 2
+            ]);
```
Can we also assert against the reservation appearing in the json?

david-fairbanks42 replied:
Yes but I'm not sure you're going to like my solution.

FatBoyXPC commented on tests/Feature/ReservationTest.php:
```diff
@@ -0,0 +1,152 @@
+
+        $response
+            ->assertStatus(200)
+            //->assertJsonStructure(['success', 'attendee'])
```
Can we delete this comment?

FatBoyXPC commented on tests/Feature/ReservationTest.php:
```diff
@@ -0,0 +1,152 @@
+            ]);
+    }
+
+    public function test_update_invalidValue()
```
Can we snake case `invalid_value`?

FatBoyXPC commented on tests/Feature/ReservationTest.php:
```diff
@@ -0,0 +1,152 @@
+            ->assertStatus(422)
+            /*->assertJson([
+                'success' => true
+            ])*/;
```
Can we clean up this assertion?

FatBoyXPC commented on tests/Feature/ReservationTest.php:
```diff
@@ -0,0 +1,152 @@
+            ->assertStatus(200)
+            ->assertJson([
+                'success' => true
+            ]);
```
Can we also assert to make sure the reservation got updated?

FatBoyXPC commented on tests/Feature/ReservationTest.php:
```diff
@@ -0,0 +1,152 @@
+            ->assertStatus(200)
+            ->assertJson([
+                'success' => true
+            ]);
```
Can we assert the reservation was deleted (in this case updated to no)?

david-fairbanks42 committed:
Changes as suggested by FatBoyXPC

Add more response tests and selected database test.
Remove related ID field in favor of whereHas method.

david-fairbanks42 commented:
These changes have been pushed and I've added responses as needed the conversation.

FatBoyXPC reviewed (approved):
LOL Totally digging "Secret Pirate Meeting" :rofl: 

:shipit:
